### PR TITLE
ddl: disallow modifying the generated expression of stored or indexed column

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve? 
A temporary solution for issue #10924 and issue #10937.

And fix issue #10973.

### What is changed and how it works?
- move the auto_increment check from ModifyColumn() into checkModifyGeneratedColumn(), which is shared by both `modify column` and `change column`, fixing issue #10973.
- refactor a tiny code block
- add a check for `indexed` or `stored` column

Now it forbids three `modify` or `change` operations:
- changing the expression of the generated `stored` column is not allowed.
- changing the expression of the generated column with indexes is not allowed.
- changing a non-generated column to a generated `stored` column is not allowed.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

Related changes

 - Need to update the documentation

### CODE REVIEW

File: ddl/db_test.go [hunk 1]
```diff
@@ -2811,6 +2811,27 @@ func (s *testDBSuite5) TestAddIndexForGeneratedColumn(c *C) {
 	s.tk.MustExec("admin check table gcai_table")
 }
 
+func (s *testDBSuite5) TestModifyIndexedGeneratedColumn(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	// Test create an exist database
+	_, err := tk.Exec("CREATE database test")
+	c.Assert(err, NotNil)
```
Comment (Deardrops) [inline ddl/db_test.go]: Why we need test the case here?
if you want to create a new database or use an exist database, try
`create database if not exists test`

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2705,6 +2705,11 @@ func (d *ddl) ModifyColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Al
 			if err := checkAutoIncrementRef(specNewColumn.Name.Name.L, dependColNames, t.Meta()); err != nil {
 				return errors.Trace(err)
 			}
+
+			// If there is an index on the generated column which we are trying to modify, return an error.
+			if tables.FindIndexByColName(t, specNewColumn.Name.Name.L) != nil {
```
Comment (crazycs520) [inline ddl/ddl_api.go]: `FindIndexByColName` is not suitable to use in here.

File: ddl/db_test.go [hunk 3]
```diff
@@ -2811,6 +2811,27 @@ func (s *testDBSuite5) TestAddIndexForGeneratedColumn(c *C) {
...
+
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1 (a int, b int as (a+1), index idx(b));")
+	tk.MustExec("insert into t1 set a=1")
+	_, err = tk.Exec("alter table t1 modify column b int as (a+2)")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[ddl:3106]'modifying an indexed column' is not supported for generated columns.")
+
+	tk.MustExec("drop index idx on t1;")
+	tk.MustExec("alter table t1 modify b int as (a + 2);")
```
Comment (Deardrops) [inline ddl/db_test.go]: I suggest add a query statement after this step to ensure the correctness of the results:
`tk.MustQuery("select * from test").Check(testkit.Rows("1 3"))`

File: ddl/ddl_api.go (hunk 2 above)
Comment (crazycs520) [reply ddl/ddl_api.go]:
```sql
mysql>create table t1 (a int, b int as (a+1), index idx(a,b));
Query OK, 0 rows affected
Time: 0.070s
mysql>insert into t1 set a=1
Query OK, 1 row affected
Time: 0.008s
mysql>alter table t1 modify column b int as (a+2)
Query OK, 0 rows affected
Time: 0.066s
mysql>select a,b from t1 ignore index (idx);
+---+---+
| a | b |
+---+---+
| 1 | 3 |
+---+---+
1 row in set
Time: 0.007s
mysql>select a,b from t1 use index (idx);
+---+---+
| a | b |
+---+---+
| 1 | 2 |
+---+---+
```

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -2705,6 +2705,18 @@ func (d *ddl) ModifyColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Al
 			if err := checkAutoIncrementRef(specNewColumn.Name.Name.L, dependColNames, t.Meta()); err != nil {
 				return errors.Trace(err)
 			}
+
+			// If there is an index on the generated column which we are trying to modify, return an error.
+			for _, idx := range t.Indices() {
+				if idx.Meta().State != model.StatePublic {
+					continue
+				}
```
Comment (tangenta) [inline ddl/ddl_api.go]: What happened if a write-only or delete-only state column is detected at this time?

File: ddl/db_test.go [hunk 5]
```diff
@@ -2811,6 +2811,33 @@ func (s *testDBSuite5) TestAddIndexForGeneratedColumn(c *C) {
...
+func (s *testDBSuite5) TestModifyIndexedGeneratedColumn(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("create database if not exists test")
+	tk.MustExec("use test")
+	modIdxColErrMsg := "[ddl:3106]'modifying an indexed column' is not supported for generated columns."
+
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1 (a int, b int as (a+1), index idx(b));")
+	tk.MustExec("insert into t1 set a=1")
+	_, err := tk.Exec("alter table t1 modify column b int as (a+2)")
```
Comment (crazycs520) [inline ddl/db_test.go]: add test for `alter table t1 modify column b bigint as (a+1)`

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -2705,6 +2705,15 @@ func (d *ddl) ModifyColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Al
 			if err := checkAutoIncrementRef(specNewColumn.Name.Name.L, dependColNames, t.Meta()); err != nil {
 				return errors.Trace(err)
 			}
+
+			// If there is an index on the generated column which we are trying to modify, return an error.
```
Comment (winkyao) [inline ddl/ddl_api.go]: Wrap this `if option.Tp == ast.ColumnOptionGenerated` code block into a function like `changeGeneratedColumns`

File: ddl/db_test.go [hunk 7]
```diff
@@ -2811,6 +2811,33 @@ func (s *testDBSuite5) TestAddIndexForGeneratedColumn(c *C) {
...
 }
 
+func (s *testDBSuite5) TestModifyIndexedGeneratedColumn(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("create database if not exists test")
+	tk.MustExec("use test")
+	modIdxColErrMsg := "[ddl:3106]'modifying an indexed column' is not supported for generated columns."
+
+	tk.MustExec("drop table if exists t1;")
+	tk.MustExec("create table t1 (a int, b int as (a+1), index idx(b));")
```
Comment (winkyao) [inline ddl/db_test.go]: What if change the expression of the stored generated column? Will it return an error?

File: ddl/generated_column.go [hunk 8]
```diff
@@ -198,6 +207,34 @@ func checkIllegalFn4GeneratedColumn(colName string, expr ast.ExprNode) error {
...
+// Check whether newColumnDef refers to any auto-increment columns.
+func checkGeneratedWithAutoInc(tableInfo *model.TableInfo, newColumnDef *ast.ColumnDef) error {
+	_, dependColNames := findDependedColumnNames(newColumnDef)
+	if err := checkAutoIncrementRef(newColumnDef.Name.Name.L, dependColNames, tableInfo); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func checkIndexOrStored(tbl table.Table, oldCol, newCol *table.Column) error {
```
Comment (winkyao) [inline ddl/generated_column.go]: One function do one thing, please check stored in another function.

File: ddl/generated_column.go [hunk 9]
```diff
@@ -198,6 +207,34 @@ func checkIllegalFn4GeneratedColumn(colName string, expr ast.ExprNode) error {
...
+		return nil
+	}
+
+	if newCol.GeneratedStored {
+		return errUnsupportedOnGeneratedColumn.GenWithStackByArgs("modifying a stored column")
+	}
+
+	for _, idx := range tbl.Indices() {
+		for _, col := range idx.Meta().Columns {
+			if col.Name.L == newCol.Name.L {
```
Comment (winkyao) [inline ddl/generated_column.go]: why not check `col.Name.L == oldCol.Name.L`?

Comment (tangenta) [reply ddl/generated_column.go]: @winkyao If the new column is NOT stored or index, then no backfilling is needed. So there is nothing to do with the old column.

File: ddl/generated_column.go (hunk 8 above)
Comment (tangenta) [reply ddl/generated_column.go]: @winkyao I agreed with you. But once backfilling is supported, we can remove the whole function entirely. Additionly, it is short and a part of logic can be reused. How do you think?

**Pull request merged**